plugins/grep/grep.7
.TH YED-GREP 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
grep \- Search for matches to regular expressions in files in your current directory.
.SH CONFIGURATION
.SS grep-prg <command>
The grep command to run where a non-backslash-escaped % is replaced with the search string.
The default value is "grep --exclude-dir={.git} -RHnIs '%' ."
.SH COMMANDS
.SS grep [regex]
If regex is provided, run the grep-prg with the regex and put results in *grep-list.
If regex is not provided, grep will be run interactively as you type the regex on the command line.
Hitting ENTER on an item in *grep-list will jump to that match.
.SH BUFFERS
*grep-list
.SH NOTES
.P
Be careful when running grep in high-level directories with many files.
For example you should probably avoid running grep in your home directory because it can take a long time to search.
.SH VERSION
0.0.1
.SH KEYWORDS
grep, search, find, regex, jump, pattern


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


plugins/grep/grep.c
#include <yed/plugin.h>

void grep(int n_args, char **args);
void grep_start(void);
void grep_cleanup(void);
void grep_take_key(int key);
void grep_run(void);
void grep_select(void);

void grep_key_pressed_handler(yed_event *event);

yed_buffer *get_or_make_buff(void) {
    yed_buffer *buff;

    buff = yed_get_buffer("*grep-list");

    if (buff == NULL) {
        buff = yed_create_buffer("*grep-list");
        buff->flags |= BUFF_RD_ONLY | BUFF_SPECIAL;
    }

    return buff;
}

static char *prg;
static char *save_current_search;

int yed_plugin_boot(yed_plugin *self) {
    yed_event_handler h;

    YED_PLUG_VERSION_CHECK();

    h.kind = EVENT_KEY_PRESSED;
    h.fn   = grep_key_pressed_handler;

    yed_plugin_add_event_handler(self, h);
    yed_plugin_set_command(self, "grep", grep);

    if (!yed_get_var("grep-prg")) {
        yed_set_var("grep-prg", "grep --exclude-dir={.git} -RHnIs '%' .");
    }

    return 0;
}

void grep(int n_args, char **args) {
    int i;
    int key;

    if (!ys->interactive_command) {
        prg = yed_get_var("grep-prg");
        if (!prg) {
            yed_cerr("'grep-prg' not set");
            return;
        }
        grep_start();
        if (n_args) {
            for (i = 0; i < strlen(args[0]); i += 1) {
                yed_cmd_line_readline_take_key(NULL, (int)args[0][i]);
            }
            array_zero_term(ys->cmd_buff);
            grep_run();
            ys->interactive_command = NULL;
            yed_clear_cmd_buff();
        }
    } else {
        sscanf(args[0], "%d", &key);
        grep_take_key(key);
    }
}

void grep_cleanup(void) {
    ys->save_search = save_current_search;
}

void grep_start(void) {
    ys->interactive_command = "grep";
    ys->cmd_prompt          = "(grep) ";
    save_current_search = ys->current_search;

    get_or_make_buff()->flags &= ~BUFF_RD_ONLY;
    yed_buff_clear_no_undo(get_or_make_buff());
    get_or_make_buff()->flags |= BUFF_RD_ONLY;

    YEXE("special-buffer-prepare-focus", "*grep-list");
    yed_frame_set_buff(ys->active_frame, get_or_make_buff());
    yed_set_cursor_far_within_frame(ys->active_frame, 1, 1);
    yed_clear_cmd_buff();
}

void grep_take_key(int key) {
    switch (key) {
        case ESC:
        case CTRL_C:
            ys->interactive_command = NULL;
            ys->current_search      = NULL;
            yed_clear_cmd_buff();
            YEXE("special-buffer-prepare-unfocus", "*grep-list");
            grep_cleanup();
            break;
        case ENTER:
            ys->interactive_command = NULL;
            ys->current_search      = NULL;
            ys->active_frame->dirty = 1;
            yed_clear_cmd_buff();
            break;
        default:
            yed_cmd_line_readline_take_key(NULL, key);
            grep_run();
            break;
    }
}

void grep_run(void) {
    char       cmd_buff[1024];
    char      *pattern;
    int        len, status;

    get_or_make_buff()->flags &= ~BUFF_RD_ONLY;

    array_zero_term(ys->cmd_buff);

    cmd_buff[0]        = 0;
    pattern            = array_data(ys->cmd_buff);
    ys->current_search = pattern;

    if (strlen(pattern) == 0)     { goto empty; }

    len = perc_subst(prg, pattern, cmd_buff, sizeof(cmd_buff));

    ASSERT(len > 0, "buff too small for perc_subst");

    strcat(cmd_buff, " 2>/dev/null");

    if (yed_read_subproc_into_buffer(cmd_buff, get_or_make_buff(), &status) != 0) {
        goto empty;
    }

    if (status != 0) {
empty:;
        yed_buff_clear_no_undo(get_or_make_buff());
    }

    get_or_make_buff()->flags |= BUFF_RD_ONLY;
}

void grep_select(void) {
    yed_line *line;
    char      _path[256];
    char     *path,
             *c;
    int       row,
              row_idx;

    path = _path;
    line = yed_buff_get_line(get_or_make_buff(), ys->active_frame->cursor_line);
    array_zero_term(line->chars);

    row_idx = 0;
    array_traverse(line->chars, c) {
        row_idx += 1;
        if (*c == ':') {
            break;
        }
        *path++ = *c;
    }
    *path = 0;
    path  = _path;

    if (row_idx >= 2) {
        if (*path == '.' && *(path + 1) == '/') {
            path += 2;
        }
    }

    sscanf(array_data(line->chars) + row_idx, "%d", &row);

    YEXE("special-buffer-prepare-jump-focus", "*grep-list");
    YEXE("buffer", path);
    yed_set_cursor_within_frame(ys->active_frame, row, 1);
    grep_cleanup();
}

void grep_key_pressed_handler(yed_event *event) {
    yed_frame *eframe;

    eframe = ys->active_frame;

    if (event->key != ENTER                           /* not the key we want */
    ||  ys->interactive_command                       /* still typing        */
    ||  !eframe                                       /* no frame            */
    ||  !eframe->buffer                               /* no buffer           */
    ||  strcmp(eframe->buffer->name, "*grep-list")) { /* not our buffer      */
        return;
    }

    grep_select();

    event->cancel = 1;
}
